src/controllers/rendering/controllerrenderingengine.cpp

```diff
@@ -0,0 +1,354 @@
+    case QImage::Format_RGB16:
+        m_GLDataFormat = GL_RGB;
+        m_GLDataType = m_screenInfo.reversedColor
+                ? GL_UNSIGNED_SHORT_5_6_5_REV
+                : GL_UNSIGNED_SHORT_5_6_5;
+        break;
+    case QImage::Format_RGB888:
+        m_GLDataFormat = m_screenInfo.reversedColor ? GL_BGR : GL_RGB;
+        m_GLDataType = GL_UNSIGNED_BYTE;
+        break;
+    case QImage::Format_RGBA8888:
+        m_GLDataFormat = m_screenInfo.reversedColor ? GL_BGRA : GL_RGBA;
+        m_GLDataType = GL_UNSIGNED_BYTE;
+        break;
+    default:
+        m_isValid = false;
+        DEBUG_ASSERT(!"Unsupported format");
+    }
+
+    if (!m_isValid) {
+        return;
+    }
+
+    prepare();
+}
+
+void ControllerRenderingEngine::prepare() {
+    m_pScheduler = std::make_unique<ControllerRenderingScheduler>(m_screenInfo);
+
+    connect(this,
+            &ControllerRenderingEngine::setupRequested,
+            this,
+            &ControllerRenderingEngine::setup);
+    connect(this,
+            &ControllerRenderingEngine::stopRequested,
+            this,
+            &ControllerRenderingEngine::finish);
+    connect(m_pScheduler.get(),
+            &ControllerRenderingScheduler::frameRequested,
+            this,
+            &ControllerRenderingEngine::renderFrame);
+
+    moveToThread(qApp->thread());
+}
+
+ControllerRenderingEngine::~ControllerRenderingEngine() {
+    VERIFY_OR_DEBUG_ASSERT(!m_fbo) {
+        kLogger.warning() << "The ControllerEngine is being deleted but hasn't been "
+                             "cleaned up. Brace for impact";
+    };
+}
+
+void ControllerRenderingEngine::start() {
+    VERIFY_OR_DEBUG_ASSERT(!m_pScheduler->isFinished() &&
+            !m_pScheduler->isInterruptionRequested()) {
+        kLogger.warning() << "Render thread has or ir about to terminate. Cannot "
+                             "start this render anymore.";
+        return;
+    }
+    m_pScheduler->start(QThread::NormalPriority);
+}
+bool ControllerRenderingEngine::isRunning() const {
+    return m_pScheduler && m_pScheduler->isRunning();
+}
+
+void ControllerRenderingEngine::requestSetup(std::shared_ptr<QQmlEngine> qmlEngine) {
+    m_isValid = false;
+    VERIFY_OR_DEBUG_ASSERT(QThread::currentThread() != thread()) {
+        kLogger.warning() << "Unable to setup OpenGL rendering context from the same "
+                             "thread as the render object";
+        return;
+    }
+    emit setupRequested(qmlEngine);
+
+    const auto lock = lockMutex(&m_mutex);
+    if (!m_quickWindow) {
+        m_waitCondition.wait(&m_mutex);
+    }
+}
+
+void ControllerRenderingEngine::requestSend(Controller* controller, const QByteArray& frame) {
+    VERIFY_OR_DEBUG_ASSERT(m_pScheduler) {
+        return;
+    }
+    m_pScheduler->requestSend(controller, frame);
+}
+
+void ControllerRenderingEngine::setup(std::shared_ptr<QQmlEngine> qmlEngine) {
+    QSurfaceFormat format;
+    format.setSamples(16);
```

Comment:
Yeah make sense, will add this to the XML spec, like for FPS

Reply:
I think it needs to be configurable per application, since it very much depends on the end users hardware whether they can afford msaa, not the mapping hardware.

Reply:
That's a nice to have, yes. But lets not forget this PR is already 4k+ and open for more than a year. I'll put that in the XML, so user can overwrite it if it unusable.